src/lib/notifier.c

```diff
@@ -14,82 +14,164 @@
 #include <sof/spinlock.h>
 #include <ipc/topology.h>
 
-static struct notify_data _notify_data __aligned(PLATFORM_DCACHE_ALIGN);
-
-void notifier_register(struct notifier *notifier)
+#define trace_notifier(__e, ...) \
+	trace_event(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+#define tracev_notifier(__e, ...) \
+	tracev_event(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+#define trace_notifier_error(__e, ...) \
+	trace_error(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+
+struct notify_data {
+	void *caller;
+	enum notify_id type;
+	uint32_t data_size;
+	void *data;
+};
+
+static struct notify_data _notify_data[PLATFORM_CORE_COUNT]
+	__aligned(PLATFORM_DCACHE_ALIGN);
+
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
+	struct list_item list;
+};
+
+void notifier_register(void *receiver, void *caller, enum notify_id type,
+		       void (*cb)(void *arg, enum notify_id type, void *data))
 {
 	struct notify *notify = *arch_notify_get();
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_register() error: invalid ID to "
+				     "register to.");
+		return;
+	}
+
+	handle = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*handle));
+
+	if (!handle) {
+		trace_notifier_error("notifier_register() error: callback "
+				     "handle allocation failed.");
+		return;
+	}
+
+	list_init(&handle->list);
+	handle->receiver = receiver;
+	handle->caller = caller;
+	handle->cb = cb;
 
 	spin_lock(notify->lock);
-	list_item_prepend(&notifier->list, &notify->list);
+	list_item_prepend(&handle->list, &notify->list[type]);
 	spin_unlock(notify->lock);
 }
 
-void notifier_unregister(struct notifier *notifier)
+void notifier_unregister(void *receiver, void *caller, enum notify_id type)
 {
 	struct notify *notify = *arch_notify_get();
+	struct list_item *wlist;
+	struct list_item *tlist;
+	struct callback_handle *handle;
+	int i;
+
+	if (type > NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_unregister() error: invalid ID "
+				     "to unregister from.");
+		return;
+	}
+
+	if (type == NOTIFIER_ID_ALL) {
+		/* recurse once for every callback type */
+		for (i = 0; i < NOTIFIER_ID_ALL; i++)
+			notifier_unregister(receiver, caller, i);
+		return;
+	}
 
 	spin_lock(notify->lock);
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
```

Comment:
Let's add comment here what is your intention meaning what is your condition in searching for the right handles.